Add docs for connecting to MetaMask (#1525)
MetaMask/metamask-docs: wallet/how-to/use-non-evm-networks/starknet-snap/index.md

---
sidebar_position: 1
sidebar_label: Starknet

> description: Learn about how Starknet Snap connects to MetaMask using `get-starknet`.
> tags:
>   - Snaps
>   - non-EVM

Reviewer: ```suggestion
description: Interact with users' Starknet accounts in MetaMask.
```

Syncing with changes I made to the skeleton PR. I think we should leave out doc tags for now, to not over-complicate things.

---

> # Build with Starknet

Reviewer: ```suggestion
# Use Starknet

You can interact with users' Starknet accounts in MetaMask by connecting to the Starknet Snap.
```

Suggested intro and title re-wording to get straight to the point.

Starknet is a Layer 2 network built on Ethereum. It processes transactions off the main Ethereum chain to increase speed and reduce costs, while still benefiting from Ethereum's security.

One major aspect of Starknet development is integrating with user wallets. There are two primary approaches to achieve this integration:

- [Using direct Snap invocation with `wallet_invokeSnap](#use-wallet_invokesnap-directly)
- [Using the `get-starknet` integration](#use-the-get-starknet-integration)

Both methods enable developers to connect Starknet dapps with wallet solutions. 

> The choice between them depends on the specific needs of the project, the desired level of control, and familiarity with Starknet and Snap technologies.

Reviewer 2: Shall we make clear the benefits of using get-starknet ? 

From get-starknet stand point they are **the wallet interoperability layer** of the starknet ecosystem. This means that a starknet dApp developer choosing get-starknet to develop will automatically be compatible with several wallet at the same time (including metamask, through the starknet-snap).

Author: I think that is mentioned later on

## About `get-starknet`

> `get-starknet` simplifies Starknet network interactions by extending MetaMask functionality. It eliminates the need for custom wallet logic and allows you to connect to Starknet and perform transactions through the MetaMask interface.

Reviewer 2: ```suggestion
`get-starknet` simplifies Starknet network interactions by extending MetaMask functionality. It eliminates the need for custom wallet logic and allows your user to connect to Starknet and perform transactions through the MetaMask interface.
```

> `get-starknet` simplifies Starknet network interactions by extending MetaMask functionality. It eliminates the need for custom wallet logic and allows you to connect to Starknet and perform transactions through the MetaMask interface.
> The `get-starknet` toolkit works with the Starknet Snap to enable interaction between dapps and the Starknet network through MetaMask. 
> Starknet Snap is a MetaMask extension designed specifically to allow users to manage Starknet accounts and perform transactions on the Starknet network. 
> When you integrate `get-starknet` into your dapp, it creates a `WalletAccount` object. `WalletAccount` acts as a connection between dapps and MetaMask and provides a structured way to manage Starknet-specific interactions. 
> This allows users to send transactions, sign messages, and manage accounts within the MetaMask interface and it can be extended to multiple wallets, with Starknet capabilities added on.

Reviewer: ```suggestion
`get-starknet` is a library that simplifies Starknet network interactions.
`get-starknet` works with the Starknet Snap to extend the functionality of MetaMask and enable dapps to interact with users' Starknet accounts in MetaMask.

When you integrate `get-starknet` into your dapp, it creates a `WalletAccount` object. `WalletAccount` acts as a connection between dapps and MetaMask and provides a structured way to manage Starknet interactions.
This allows users to send Starknet transactions, sign Starknet messages, and manage Starknet accounts within MetaMask, and this functionality can be extended to multiple wallets.
```

Suggest re-wording – this page needs to be shortened in general. Also let's use one term consistently to refer to the `get-starknet`: tool, toolkit, library, integration, etc?

### How `get-starknet` and MetaMask interact

> The app uses `get-starknet` to request a connection to MetaMask. If the Starknet Snap isn't installed, MetaMask prompts the user to connect and approve the installation. 
> After it is approved, `get-starknet` receives a Starknet Windows Object (SWO), which represents the MetaMask wallet with Starknet functionality.
> After receiving the SWO, `get-starknet` creates a WalletAccount instance. 
> This instance manages the Starknet account within MetaMask. With this setup, `get-starknet` enables the app to use Starknet features through MetaMask with the following capabilities:
> 
> - `WalletAccount` uses a specified provider to read data from the Starknet network.
> - For transactions, `get-starknet` prepares the transaction data and sends it to MetaMask for signing through the Starknet Snap.
> - `get-starknet` allows the dapp to create contract instances connected to the `WalletAccount`. These instances can invoke smart contract functions, with MetaMask handling the necessary signatures.
> - `get-starknet` sets up listeners for account and network changes within MetaMask.
> The dapp can subscribe to these events to update its state accordingly.
> - `get-starknet` can request network changes through MetaMask, and allows users to switch between different Starknet networks (for example Mainnet, testnet).
> - `get-starknet` can request MetaMask to display specific tokens in its interface, enhancing user experience.
> 

Reviewer: ```suggestion
A dapp with `get-starknet` installed interacts with MetaMask as follows:

1. The dapp uses `get-starknet` to request the user connect to MetaMask. If the user doesn't have the Starknet Snap installed, MetaMask prompts the user to connect and approve the installation.

2. Once the dapp is connected to MetaMask and the Starknet Snap, `get-starknet` receives a Starknet Windows Object (SWO), which represents the MetaMask wallet with Starknet functionality.

3. `get-starknet` creates a WalletAccount instance.
This instance manages the Starknet account within MetaMask.

With this setup, `get-starknet` enables the dapp to use Starknet features through MetaMask with the following capabilities:

- `WalletAccount` uses a specified provider to read data from the Starknet network.
- For transactions, `get-starknet` prepares the transaction data and sends it to MetaMask for signing through the Starknet Snap.
- `get-starknet` allows the dapp to create contract instances connected to the `WalletAccount`. These instances can invoke smart contract functions, with MetaMask handling the necessary signatures.
- `get-starknet` sets up listeners for account and network changes within MetaMask.
The dapp can subscribe to these events to update its state accordingly.
- `get-starknet` can request network changes through MetaMask, and allows users to switch between different Starknet networks (for example Mainnet, testnet).
- `get-starknet` can request MetaMask to display specific tokens in its interface, enhancing user experience.
```

Maybe add the flow in ordered steps?

    get->>mm: Request signature
    mm->>Snap: Sign transaction
    Snap-->>mm: Signed transaction
    mm-->>get: Return signature
    get->>network: Submit transaction
    network-->>get: Transaction result
    get-->>dapp: Transaction confirmation
    mm->>get: Account/Network changed
    get->>dapp: Notify change
```

## Compare wallet integration methods for Starknet

You can use `wallet_invokeSnap` or `get-starknet` to integrate with Starknet dapps.
The choice depends on your specific use case and development preferences. 

| Feature | `wallet_invokeSnap` | `get-starknet` |
|---------|---------------------|----------------|
| API Level | Low-level access to Starknet Snap methods | High-level abstractions for Starknet operations |
| Wallet Support | MetaMask only | Multiple Starknet-compatible wallets |
| Functionality | Limited to Snap-implemented methods | Comprehensive toolkit for Starknet dapp development |
| Multi-wallet Sync | No | Yes |
| Standardization | Might lead to non-standard implementations | Enforces standardized wallet interactions |
| Flexibility | High, with direct access to all Snap features | Moderate, within implemented features |

> | Code Portability | Low, tied to MetaMask | High, works across multiple wallet implementations |

Reviewer: I think move this table higher up on this page so users can immediately compare the actual features of each method. Shorten up the intro and add this there instead.

### Use the `get-starknet` integration

With `get-starknet`, the process is simplified:

1. Connect to a wallet using `get-starknet`.
2. Use the provided high-level methods to perform Starknet operations.

Example for sending the same transaction:

```javascript "title="starknetSnapInteraction.js""
import { connect } from 'get-starknet';

async function sendTransaction() {
  try {
    const starknet = await connect();
    if (!starknet) throw new Error('Failed to connect to wallet');

    const result = await starknet.account.execute({
      contractAddress: '0x123...',
      entrypoint: 'transfer',
      calldata: ['0x456...', '1000']
    });
    console.log('Transaction hash:', result.transaction_hash);
  } catch (error) {
    console.error('Error:', error);
  }
}
```

`get-starknet` provides a high-level, intuitive API that abstracts complex operations and manages wallet connections. 

> It offers multi-wallet support, standardized error handling, and more readable code compared to direct Snap invocation so you are not required to manage lower-level Starknet interactions.

Reviewer: This page should be a short, high-level intro to Starknet. These instructions should be moved to the existing "Connect to Starknet" guide – you can add the `wallet_invokeSnap` instructions underneath the `get-starknet` section.

Author: I was trying to highlight/compare the actions after the app is developed. I don''t think this belongs in the connect section. Maybe in manage, but the point was to show the benefits using `get-starknet. Maybe I'll just remove this completely
